theforeman.org/pipelines/lib/packaging.groovy

```diff
@@ -63,3 +108,302 @@ def repoclosures(repo, releases, version) {
 
     return closures
 }
+
+def build_deb_package_steps(packages_to_build, version, repoowner = 'theforeman', pull_request = false) {
+    def steps = [
+        setup_sources: [:],
+        executes: [:],
+        stage_packages: [:]
+    ]
+
+    for(int i = 0; i < packages_to_build.size(); i++) {
+        def index = i
+        def pkg_to_build = packages_to_build[index]
+        def project = pkg_to_build['name']
+
+        echo "${pkg_to_build}"
+
+        for(int j = 0; j < pkg_to_build['operating_systems'].size(); j++) {
+            def os_index = j
+            def os = pkg_to_build['operating_systems'][j]
+            def execute_pbuilder_dir
+
+            steps.setup_sources["${os}-${project}"] = {
+                execute_pbuilder_dir = setup_sources_deb(pkg_to_build['type'], project, os, version, repoowner, pull_request)
+            }
+
+            steps.executes["${os}-${project}"] = {
+                execute_pbuilder(execute_pbuilder_dir, os, version)
+            }
+
+            steps.stage_packages["${os}-${project}"] = {
+                stage_debian_packages(pkg_to_build['type'], project, os, version, repoowner, pull_request)
+            }
+        }
+    }
+
+    return steps
+}
+
+def deb_build_dir(type, project, os = null) {
+    def build_dir
+
+    if (type == 'core') {
+        build_dir = "debian/${os}/build-${project}"
+    } else if (type == 'dependency') {
+        build_dir = "dependencies/${os}/build-${project}"
+    } else if (type == 'plugin') {
+        build_dir = "plugins/build-${project}"
+    } else {
+        echo "Unsupported type specified: ${type}"
+        exit 1
+    }
+
+    return build_dir
+}
+
+def setup_sources_deb(type, project, os, version, repoowner, pull_request) {
+
+    if (type == 'core') {
+        setup_sources_core(project, os, version, repoowner, pull_request)
+    } else if (type == 'dependency') {
+        setup_sources_dependency(project, os, version, repoowner, pull_request)
+    } else if (type == 'plugin') {
+        setup_sources_plugin(project, os, version, repoowner, pull_request)
+    } else {
+        echo "Unsupported type specified: ${type}"
+        exit 1
+    }
+
+}
+
+def setup_sources_core(os, project, version, repoowner, pull_request = false) {
+    def package_version = sh(returnStdout: true, script: "head -n1 debian/${os}/${project}/changelog|awk '{print \$2}'|sed 's/(//;s/)//'|cut -f1 -d-").trim()
+    def build_dir = deb_build_dir('core', project, os)
+
+    package_version = package_version.replaceAll('~rc', '-rc')
+    last_commit = package_version
+
+    sh "mkdir -p ${build_dir}"
+
+    dir(build_dir) {
+        if (version =~ 'develop') {
+            def url_base = 'https://ci.theforeman.org'
+            def job_name = nightly_jenkins_job
+            def job_id = nightly_jenkins_job_id
+            def job_url = "${url_base}/job/${job_name}/${job_id}"
+
+            sh """
+              wget "${job_url}/artifact/*zip*/archive.zip"
```

Comment:
I'm wondering if we could use a native method like [copyArtifacts](https://jenkinsci.github.io/job-dsl-plugin/#method/javaposse.jobdsl.dsl.helpers.step.StepContext.copyArtifacts).